crates/red_knot_python_semantic/src/types.rs

```diff
@@ -151,6 +165,73 @@ pub(crate) fn definitions_ty<'db>(
     }
 }
 
+/// Build a declared type from a [`DeclarationsIterator`].
+///
+/// Will return a union if there is more than one declaration, or at least one plus the possibility
+/// of undeclared.
+///
+/// If undeclared is a possibility, `Unknown` type will be part of the return type.
+///
+/// # Panics
+/// Will panic if there are no declarations and no possibility of undeclared. This is a logic
+/// error, as any symbol with zero live declarations clearly may (must!) be undeclared.
+fn declarations_ty<'db>(
+    db: &'db dyn Db,
+    declarations: DeclarationsIterator<'_, 'db>,
+) -> DeclaredType<'db> {
+    let may_be_undeclared = declarations.may_be_undeclared();
+    let decl_types = declarations.map(|declaration| declaration_ty(db, declaration));
+
+    let mut all_types = (if may_be_undeclared {
+        Some(Type::Unknown)
+    } else {
+        None
+    })
+    .into_iter()
+    .chain(decl_types);
+
+    let first = all_types.next().expect(
+        "declarations_ty must not be called with zero declarations and no may-be-undeclared.",
+    );
+
+    let mut conflicting: Vec<Type<'db>> = vec![];
+    let declared_ty = if let Some(second) = all_types.next() {
+        let mut builder = UnionBuilder::new(db).add(first);
+        for other in [second].into_iter().chain(all_types) {
+            if !first.is_equivalent_to(db, other) {
+                conflicting.push(other);
+            }
+            builder = builder.add(other);
```

Comment:
I guess it's kind of difficult with the current builder API but isn't `!first.is_equivalent_to` the same as testing if the item was added to the builder?

Reply:
Right now the builder doesn't do any de-duplication. We could easily add de-duplication based on equivalence, but that would just be a temporary placeholder, as really it should be based on subtyping, e.g. if `class A(B): ...` then adding `A` to a union that has `B` in it already should not add anything to the union. But `A` and `B` are still conflicting type declarations. So ultimately this check should not be the same as union de-duplication.

Reply:
The other alternative is to use `itertools::join`. But that has the downside that it allocates a string (but that's probably fine here)

Reply:
> The other alternative is to use `itertools::join`. But that has the downside that it allocates a string (but that's probably fine here)

I think this comment was meant for a different thread? I went ahead and implemented `FormatterJoinExtension` trait rather than using `itertools::join`.